ibm/data_source_ibm_is_security_group_rules.go

```diff
@@ -0,0 +1,275 @@
+	if err != nil {
+		return err
+	}
+
+	listSecurityGroupRuleOptions := vpcv1.ListSecurityGroupRulesOptions{
+		SecurityGroupID: &secGrpId,
+	}
+
+	ruleList, response, err := sess.ListSecurityGroupRules(&listSecurityGroupRuleOptions)
+	if err != nil {
+		return fmt.Errorf("Error fetching security group rules %s\n%s", err, response)
+	}
+
+	rulesInfo := make([]map[string]interface{}, 0)
+	for _, rule := range ruleList.Rules {
+		l := map[string]interface{}{}
+		switch reflect.TypeOf(rule).String() {
+		case "*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolAll":
+			{
+				rulex := rule.(*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolAll)
+				l["direction"] = *rulex.Direction
+				l["href"] = *rulex.Href
+				l["id"] = *rulex.ID
+				l["ip_version"] = *rulex.IPVersion
+				l["protocol"] = *rulex.Protocol
+				// nested map for remote.
+				if rulex.Remote != nil {
+					remoteList := []map[string]interface{}{}
+					remoteMap := dataSourceSecurityGroupRuleCollectionRulesRemoteToMap(*rulex.Remote.(*vpcv1.SecurityGroupRuleRemote))
+					remoteList = append(remoteList, remoteMap)
+					l["remote"] = remoteList
+				}
+
+			}
+		case "*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolIcmp":
+			{
+				rulex := rule.(*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolIcmp)
+				l["direction"] = *rulex.Direction
+				l["href"] = *rulex.Href
+				l["id"] = *rulex.ID
+				l["ip_version"] = *rulex.IPVersion
+				l["code"] = *rulex.Code
+				l["protocol"] = *rulex.Protocol
+				l["type"] = *rulex.Type
+				// remote
+				if rulex.Remote != nil {
+					remoteList := []map[string]interface{}{}
+					remoteMap := dataSourceSecurityGroupRuleCollectionRulesRemoteToMap(*rulex.Remote.(*vpcv1.SecurityGroupRuleRemote))
+					remoteList = append(remoteList, remoteMap)
+					l["remote"] = remoteList
+				}
+			}
+		case "*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolTcpudp":
+			{
+				rulex := rule.(*vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolTcpudp)
+				l["direction"] = *rulex.Direction
+				l["href"] = *rulex.Href
+				l["id"] = *rulex.ID
+				l["ip_version"] = *rulex.IPVersion
+				l["protocol"] = *rulex.Protocol
+				l["port_max"] = *rulex.PortMax
+				l["port_min"] = *rulex.PortMin
+				// remote
+				if rulex.Remote != nil {
+					remoteList := []map[string]interface{}{}
+					remoteMap := dataSourceSecurityGroupRuleCollectionRulesRemoteToMap(*rulex.Remote.(*vpcv1.SecurityGroupRuleRemote))
+					remoteList = append(remoteList, remoteMap)
+					l["remote"] = remoteList
+				}
+			}
+		}
+		rulesInfo = append(rulesInfo, l)
+	}
+	d.SetId(dataSourceIBMIsSecurityGroupRulesID(d))
+	d.Set("rules", rulesInfo)
+	return nil
+}
+
+// dataSourceIBMIsSecurityGroupRulesID returns a reasonable ID for the list.
+func dataSourceIBMIsSecurityGroupRulesID(d *schema.ResourceData) string {
+	return time.Now().UTC().String()
+}
+
+func dataSourceSecurityGroupRuleCollectionRulesRemoteToMap(remoteItem vpcv1.SecurityGroupRuleRemote) (remoteMap map[string]interface{}) {
```

Comment:
can we re-use dataSourceSecurityGroupRuleRemoteToMap from ibm/data_source_ibm_is_security_group_rule.go
if both does the same thing?

Reply:
done.